apps/emqx_rule_engine/src/emqx_rule_utils.erl

```diff
@@ -87,6 +91,22 @@ preproc_tmpl([[Str, Phld]| Tokens], Acc) ->
 preproc_tmpl([[Str]| Tokens], Acc) ->
     preproc_tmpl(Tokens, put_head(str, Str, Acc)).
 
+%% Replace a string contains vars to another string in which the placeholders are replace by the
+%% corresponding values. For example, given "a: ${var}", if the var=1, the result string will be:
+%% "a: 1".
+replace_vars_in_str(Tokens, Data) when is_list(Tokens) ->
```

Comment:
Couldn't be the names a bit misleading 🤔 The replacement ocurrs not in a string but in a (maybe) preprocessed template, and if we actually pass a string `replace_vars_in_str("a: ${var}", ...)` probably a clause error will occur somewhere in `proc_tmpl`

Reply:
So could we change the name to `replace_vars_to_str`? This function is to convert tokens to a string (binary()).